be/src/exprs/agg/group_concat.h

```diff
@@ -292,4 +299,398 @@ class GroupConcatAggregateFunction
+    // update without null elements
+    void update(FunctionContext* ctx, const Column& column, size_t index, size_t offset, size_t count) {
+        (*data_columns)[index]->append(column, offset, count);
+    }
+
+    // order-by items may be null
+    void update_nulls(FunctionContext* ctx, size_t index, size_t count) {
+        DCHECK((*data_columns)[index]->is_nullable());
+        (*data_columns)[index]->append_nulls(count);
+    }
+
+    // release the trailing order-by columns
+    void release_order_by_columns() const {
+        DCHECK(data_columns != nullptr);
+        for (auto i = output_col_num + 1; i < data_columns->size(); ++i) { // after the separator column
+            data_columns->at(i).reset();
+        }
+        data_columns->resize(output_col_num + 1);
+    }
+
+    ~GroupConcatAggregateStateV2() {
+        if (data_columns != nullptr) {
+            for (auto& col : *data_columns) {
+                col.reset();
+            }
+            data_columns->clear();
+            delete data_columns;
+            data_columns = nullptr;
+        }
+    }
+    // using pointer rather than vector to avoid variadic size
+    // group_concat(a, b order by c, d), the a,b,',',c,d are put into data_columns in order, and reject null for
+    // output columns a and b.
+    Columns* data_columns = nullptr;
+    int output_col_num = 0;
+};
+
+// group_concat concatenates non-null values from a group, and output null if the group is empty.
+// TODO(fzh) we can further optimize group_concat in following 3 ways:
+// 1. reuse columns for order-by clause, group_concat(a order by 1) can avoid replacing '1' with a in plan, just keep a;
+// 2. convert output columns to string in finalized phase, instead of add cast functions in plan, which leads to
+// redundancy columns in intermediate results. For example, group_concat(a,b order by 1,2) is rewritten to
+// group_concat(cast(a to string), cast(b to string) order by a, b), resulting to keeping 4 columns, but it only needs
+// keep 2 columns in intermediate results.
+// 3. refactor order-by and distinct function to a combinator to clean the code.
+class GroupConcatAggregateFunctionV2
+        : public AggregateFunctionBatchHelper<GroupConcatAggregateStateV2, GroupConcatAggregateFunctionV2> {
+public:
+    // group_concat(a, b order by c, d), the arguments are a,b,',',c,d
+    void create_impl(FunctionContext* ctx, GroupConcatAggregateStateV2& state) const {
+        auto num = ctx->get_num_args();
+        state.data_columns = new Columns;
+        auto order_by_num = ctx->get_nulls_first().size();
+        state.output_col_num = num - order_by_num - 1; // excluding separator column
+        DCHECK(state.output_col_num > 0);
+        for (auto i = 0; i < state.output_col_num; ++i) {
+            if (UNLIKELY(!is_string_type(ctx->get_arg_type(i)->type))) {
+                ctx->set_error(fmt::format("{}-th input of group_concat should be string type.", i + 1).c_str(), false);
```

Comment:
What's the behavior of this? this should not check here?

Reply:
safety check at create, if error, the agg will report error and stop.